File: 2352. Equal Row and Column Pairs/main.cpp
#include <map>
#include <vector>

using namespace std;

class Solution {
public:
    int equalPairs(vector<vector<int>> &grid) {
        int rowSize = grid.size();
        int colSize = grid[0].size();
        int count = 0;

        for (size_t c = 0; c < colSize; c++) {
            vector<int> column;
            for (size_t r = 0; r < rowSize; r++) {
                column.push_back(grid[r][c]);
            }

            for (size_t r = 0; r < rowSize; r++) {
                if (grid[r] == column) {
                    count++;
                }
            }
        }

        return count;
    }
};

class Solution {
public:
    int equalPairs(vector<vector<int>> &grid) {
        int rowSize = grid.size();
        int colSize = grid[0].size();
        int count = 0;
        map<vector<int>, int> maps;

        for (size_t i = 0; i < rowSize; i++) {
            maps[grid[i]]++;
        }

        for (size_t j = 0; j < colSize; j++) {
            vector<int> column;
            for (size_t i = 0; i < rowSize; i++) {
                column.push_back(grid[i][j]);
            }

            if (maps.contains(column)) {
                count += maps[column];
            }
        }

        return count;
    }
};

class Solution {
    static constexpr int PRIME_BASE = 301;
    static constexpr int PRIME_MOD = 100001;

public:
    int equalPairs(vector<vector<int>> &grid) {
        int rowSize = grid.size();
        int colSize = grid[0].size();
        vector<int> rowHash(rowSize, 0);
        vector<int> colHash(colSize, 0);

        for (size_t r = 0; r < rowSize; r++) {
            int hash = 0;
            for (size_t c = 0; c < colSize; c++) {
                hash = (hash * PRIME_BASE + grid[r][c]) % PRIME_MOD;
            }
            rowHash[r] = hash;
        }

        for (size_t c = 0; c < colSize; c++) {
            int hash = 0;
            for (size_t r = 0; r < rowSize; r++) {
                hash = (hash * PRIME_BASE + grid[r][c]) % PRIME_MOD;
            }
            colHash[c] = hash;
        }

        int count = 0;

        for (size_t r = 0; r < rowSize; r++) {
            for (size_t c = 0; c < colSize; c++) {
                if (rowHash[r] == colHash[c]) {
                    count++;
                }
            }
        }

        return count;
    }
};

File: 2352. Equal Row and Column Pairs/main.py
from collections import defaultdict
from typing import List


class Solution:
    def equalPairs(self, grid: List[List[int]]) -> int:
        rowSize = len(grid)
        colSize = len(grid[0])
        count = 0

        for c in range(colSize):
            column = [grid[r][c] for r in range(rowSize)]

            for r in range(rowSize):
                if grid[r] == column:
                    count += 1

        return count


class Solution:
    def equalPairs(self, grid: List[List[int]]) -> int:
        PRIME_BASE = 301
        PRIME_MOD = 100001

        rowSize = len(grid)
        colSize = len(grid[0])
        rowHash = []
        colHash = []

        for r in range(rowSize):
            hash = 0
            for c in range(colSize):
                hash = (hash * PRIME_BASE + grid[r][c]) % PRIME_MOD
            rowHash.append(hash)

        for c in range(colSize):
            hash = 0
            for r in range(rowSize):
                hash = (hash * PRIME_BASE + grid[r][c]) % PRIME_MOD
            colHash.append(hash)

        count = 0

        for i in range(len(rowHash)):
            for j in range(len(colHash)):
                if rowHash[i] == colHash[j]:
                    count += 1

        return count


class Solution:
    def equalPairs(self, grid: List[List[int]]) -> int:
        rowSize = len(grid)
        colSize = len(grid[0])
        maps = defaultdict(int)
        count = 0

        for r in range(rowSize):
            maps[tuple(grid[r])] += 1

        for c in range(colSize):
            column = []
            for r in range(rowSize):
                column.append(grid[r][c])

            if tuple(column) in maps:
                count += maps[tuple(column)]

        return count
